presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java

```diff
@@ -72,50 +81,73 @@ private RetryDriver()
 
     public final RetryDriver onRetry(Runnable retryRunnable)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, Optional.ofNullable(retryRunnable), stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, Optional.ofNullable(retryRunnable), stopRetrying, classifier, retryExecutorService);
     }
 
     public RetryDriver stopRetryingWhen(Predicate<Exception> stopRetrying)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
     }
 
     public RetryDriver withClassifier(Function<Exception, Exception> classifier)
     {
-        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier);
+        return new RetryDriver(maxAttempts, minSleepTime, maxSleepTime, scaleFactor, maxRetryTime, retryRunnable, stopRetrying, classifier, retryExecutorService);
+    }
+
+    public <V> ListenableFuture<V> runAsync(String callableName, Callable<ListenableFuture<V>> callable)
+    {
+        requireNonNull(callableName, "callableName is null");
+        requireNonNull(callable, "callable is null");
+        return runAsyncInternal(callableName, callable, new RetryStatus());
     }
 
     public <V> V run(String callableName, Callable<V> callable)
     {
         requireNonNull(callableName, "callableName is null");
         requireNonNull(callable, "callable is null");
+        return runInternal(callableName, callable, new RetryStatus());
+    }
 
-        long startTime = System.nanoTime();
-        int attempt = 0;
+    private <V> ListenableFuture<V> runAsyncInternal(String callableName, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
+    {
+        ListenableFuture<V> resultFuture = runInternal(callableName, callable, retryStatus);
+        return Futures.catchingAsync(resultFuture, Exception.class, e -> {
+            if (shouldStopRetry(e, retryStatus)) {
+                throw propagate(e);
+            }
+            log.warn("Failed on executing %s with attempt %d, will perform async retry. Exception: %s", callableName, retryStatus.getAttempts(), e.toString());
+            long delayInMs = getRetryDelayInMs(retryStatus);
+            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, callable, retryStatus), delayInMs, TimeUnit.MILLISECONDS));
+        });
+    }
+
+    private <V> V runInternal(String callableName, Callable<V> callable, RetryStatus retryStatus)
+    {
         while (true) {
-            attempt++;
+            retryStatus.nextAttempt();
 
-            if (attempt > 1) {
+            if (retryStatus.getAttempts() > 1) {
```

Comment:
Since you use AtomitInteger, you can merge these two calls in one and use `incrementAndGet`

Reply:
Done.